<a href="https://colab.research.google.com/github/MochamadPhilliaWibowo/Machine-Learning-Class/blob/main/Tugas%20Week%2013/BAB%201/Transformers%2C_what_can_they_do_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nama : Mochamad Phillia Wibowo

NIM : 1103204191

Kelas : Machine Learning (TK-44-G04)

Tugas 13 NLP HUGGING FACE (BAB 1)

# Transformers, what can they do?

Instalasi the Transformers, Datasets, dan Evaluasi library untuk menjalankan program ini.

In [1]:
!pip install datasets evaluate transformers[sentencepiece]

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 542.1/542.1 kB 5.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 6.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 8.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.9/64.9 kB 4.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 7.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 12.9 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.31.0
    Uninstalling requests-2.31.0:
      Successfully uninstalled requests-2.31.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires requests==2.31.0, but you have requests 2.32.3 which is incompatible.


Ini akan menginstal semua pustaka yang diperlukan untuk pengembangan NLP dan pemrosesan bahasa alami dengan menggunakan Python.

Pustaka-pustaka telah berhasil diinstal. Namun, perhatikan bahwa terjadi konflik versi antara pustaka requests yang dibutuhkan oleh google-colab dengan versi yang diinstal oleh datasets. Hal ini dapat memengaruhi fungsionalitas yang bergantung pada pustaka requests.

In [2]:
from transformers import pipeline

classifier = pipeline("sentiment-analysis")
classifier("I've been waiting for a HuggingFace course my whole life.")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that aut

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

[{'label': 'POSITIVE', 'score': 0.9598048329353333}]

Code di atas menggunakan Transformers untuk membuat sebuah pipeline untuk analisis sentimen dalam bahasa Inggris. Berikut langkah-langkahnya:

1. from transformers import pipeline: Ini mengimpor fungsi pipeline dari pustaka Transformers.

2. classifier = pipeline("sentiment-analysis"): Ini membuat sebuah pipeline untuk analisis sentimen. Dengan menggunakan string "sentiment-analysis" sebagai argumen, pipeline akan menggunakan model yang telah dilatih sebelumnya untuk melakukan analisis sentimen.

3. classifier("I've been waiting for a HuggingFace course my whole life."): Ini mengambil teks sebagai masukan dan menggunakan model yang telah dilatih sebelumnya untuk menganalisis sentimen teks tersebut. Hasilnya adalah label sentimen (misalnya, "POSITIVE" atau "NEGATIVE") dan skor sentimen yang menunjukkan seberapa kuatnya sentimen tersebut dalam teks yang diberikan.

In [3]:
classifier(
    ["I've been waiting for a HuggingFace course my whole life.", "I hate this so much!"]
)

[{'label': 'POSITIVE', 'score': 0.9598048329353333},
 {'label': 'NEGATIVE', 'score': 0.9994558691978455}]

Code tersebut memanfaatkan model yang sudah dilatih untuk melakukan analisis sentimen pada dua teks yang diberikan secara bersamaan. Dengan menyediakan list teks sebagai input, kita dapat menganalisis sentimen dari masing-masing teks dalam satu kali pemanggilan fungsi. Berikut adalah langkah-langkahnya:

1. "I've been waiting for a HuggingFace course my whole life.": Ini adalah teks pertama yang akan dianalisis sentimennya. Teks ini mungkin akan dianggap positif karena menyatakan antusiasme terhadap kursus HuggingFace.

2. "I hate this so much!": Ini adalah teks kedua yang akan dianalisis sentimennya. Teks ini mungkin akan dianggap negatif karena menyatakan kebencian.

Dengan menggunakan classifier yang sudah didefinisikan sebelumnya, kita dapat langsung memanggilnya dengan list teks tersebut sebagai argumen, dan hasilnya akan berupa list label sentimen dan skor sentimen untuk masing-masing teks dalam list tersebut.

Hasil output tersebut menunjukkan analisis sentimen untuk dua teks yang diberikan:

Teks pertama ("I've been waiting for a HuggingFace course my whole life.") memiliki label sentimen "POSITIVE" dengan skor sentimen sebesar 0.9598. Ini menandakan bahwa teks tersebut memiliki sentimen positif.

Teks kedua ("I hate this so much!") memiliki label sentimen "NEGATIVE" dengan skor sentimen sebesar 0.9995. Ini menandakan bahwa teks tersebut memiliki sentimen negatif dengan tingkat kepercayaan yang sangat tinggi.

In [4]:
from transformers import pipeline

classifier = pipeline("zero-shot-classification")
classifier(
    "This is a course about the Transformers library",
    candidate_labels=["education", "politics", "business"],
)

No model was supplied, defaulted to facebook/bart-large-mnli and revision c626438 (https://huggingface.co/facebook/bart-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/1.15k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

{'sequence': 'This is a course about the Transformers library',
 'labels': ['education', 'business', 'politics'],
 'scores': [0.8445993661880493, 0.1119738519191742, 0.04342673718929291]}

Code tersebut menggunakan model zero-shot classification dari Transformers untuk melakukan klasifikasi teks tanpa memerlukan label kelas tertentu. Berikut adalah detailnya:

1. "This is a course about the Transformers library": Ini adalah teks yang akan diklasifikasikan.

2. candidate_labels=["education", "politics", "business"]: Ini adalah daftar label kandidat yang mungkin cocok dengan teks yang diberikan. Dalam hal ini, label kandidat adalah "education" (pendidikan), "politics" (politik), dan "business" (bisnis).

Dengan menggunakan classifier yang telah didefinisikan sebelumnya, kita memanggilnya dengan teks yang akan diklasifikasikan dan daftar label kandidat sebagai argumen. Hasilnya adalah label yang dipilih untuk teks tersebut berdasarkan kandidat label yang diberikan.

Output tersebut memberikan hasil klasifikasi zero-shot untuk teks yang diberikan:

- sequence: Ini adalah teks yang diberikan sebagai input.

- labels: Ini adalah label-label yang diberikan sebagai kandidat kelas yang mungkin cocok dengan teks yang diberikan. Urutan label ini sesuai dengan urutan skor yang terkait.

- scores: Ini adalah skor yang menunjukkan seberapa cocok setiap label dengan teks yang diberikan. Semakin tinggi skor, semakin cocok label tersebut dengan teks yang diberikan. Dalam hal ini, label "education" memiliki skor tertinggi, diikuti oleh "business", dan "politics" memiliki skor terendah.

In [5]:
from transformers import pipeline

generator = pipeline("text-generation")
generator("In this course, we will teach you how to")

No model was supplied, defaulted to openai-community/gpt2 and revision 6c0e608 (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': "In this course, we will teach you how to make coffee from scratch in less than 90 minutes instead of a week's time. You will learn how to write and format your coffee, develop your skills using the same tools you use on the job right"}]

Code tersebut menggunakan model generasi teks dari Transformers untuk menghasilkan teks berlanjut dari awal yang diberikan. Berikut adalah detailnya:

"In this course, we will teach you how to": Ini adalah teks awal yang akan digunakan sebagai dasar untuk menghasilkan teks lanjutan.
Dengan menggunakan generator yang telah didefinisikan sebelumnya, kita memanggilnya dengan teks awal yang diberikan sebagai argumen. Hasilnya adalah teks yang dihasilkan berdasarkan konteks awal yang diberikan.

In [6]:
from transformers import pipeline

generator = pipeline("text-generation", model="distilgpt2")
generator(
    "In this course, we will teach you how to",
    max_length=30,
    num_return_sequences=2,
)

config.json:   0%|          | 0.00/762 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/353M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'In this course, we will teach you how to write a JavaScript code and get you into Ruby, JavaScript, and Java.\n\n\n1.'},
 {'generated_text': 'In this course, we will teach you how to apply an application of both the language of language and the concept of translation. In this course, we'}]

Code tersebut menggunakan model generasi teks GPT-2 dari Transformers untuk menghasilkan multiple teks berlanjut dari awal yang diberikan. Berikut adalah detailnya:

1. "In this course, we will teach you how to": Ini adalah teks awal yang akan digunakan sebagai dasar untuk menghasilkan teks lanjutan.

2. max_length=30: Parameter ini menentukan panjang maksimum dari teks yang dihasilkan. Dalam hal ini, maksimal 30 token akan dihasilkan.

3. num_return_sequences=2: Parameter ini menentukan jumlah teks yang akan dihasilkan. Dalam hal ini, dua teks akan dihasilkan.

Dengan menggunakan generator yang telah didefinisikan sebelumnya, kita memanggilnya dengan teks awal dan parameter-parameter yang diberikan. Hasilnya adalah dua teks yang dihasilkan berdasarkan konteks awal yang diberikan.

In [7]:
from transformers import pipeline

unmasker = pipeline("fill-mask")
unmasker("This course will teach you all about <mask> models.", top_k=2)

No model was supplied, defaulted to distilbert/distilroberta-base and revision ec58a5b (https://huggingface.co/distilbert/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/480 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/331M [00:00<?, ?B/s]

Some weights of the model checkpoint at distilbert/distilroberta-base were not used when initializing RobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

[{'score': 0.19619794189929962,
  'token': 30412,
  'token_str': ' mathematical',
  'sequence': 'This course will teach you all about mathematical models.'},
 {'score': 0.04052729159593582,
  'token': 38163,
  'token_str': ' computational',
  'sequence': 'This course will teach you all about computational models.'}]

Code tersebut menggunakan model BERT dari Transformers untuk melakukan pengisian kata tersembunyi (fill-mask) dalam sebuah teks. Berikut adalah detailnya:

1. "This course will teach you all about <mask> models.": Ini adalah teks yang mengandung sebuah tempat untuk pengisian kata tersembunyi (mask). Di sini, <mask> adalah tempat untuk mengisi kata yang hilang.

2. top_k=2: Parameter ini menentukan jumlah kandidat kata yang paling mungkin untuk mengisi tempat yang tersembunyi dalam teks. Dalam hal ini, dua kandidat kata yang paling mungkin akan ditampilkan.

Dengan menggunakan unmasker yang telah didefinisikan sebelumnya, kita memanggilnya dengan teks yang mengandung tempat untuk pengisian kata tersembunyi dan parameter-parameter yang diberikan. Hasilnya adalah dua kandidat kata yang paling mungkin untuk mengisi tempat yang tersembunyi dalam teks tersebut.

In [8]:
from transformers import pipeline

ner = pipeline("ner", grouped_entities=True)
ner("My name is Sylvain and I work at Hugging Face in Brooklyn.")

No model was supplied, defaulted to dbmdz/bert-large-cased-finetuned-conll03-english and revision f2482bf (https://huggingface.co/dbmdz/bert-large-cased-finetuned-conll03-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/998 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.33G [00:00<?, ?B/s]

Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/60.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/pipelines/token_classification.py:168: UserWarning: `grouped_entities` is deprecated and will be removed in version v5.0.0, defaulted to `aggregation_strategy="simple"` instead.
  warnings.warn(


[{'entity_group': 'PER',
  'score': 0.9981694,
  'word': 'Sylvain',
  'start': 11,
  'end': 18},
 {'entity_group': 'ORG',
  'score': 0.9796019,
  'word': 'Hugging Face',
  'start': 33,
  'end': 45},
 {'entity_group': 'LOC',
  'score': 0.9932106,
  'word': 'Brooklyn',
  'start': 49,
  'end': 57}]

Code tersebut menggunakan model NER (Named Entity Recognition) dari Transformers untuk menemukan entitas berupa nama dalam sebuah teks. Berikut adalah detailnya:

1. "My name is Sylvain and I work at Hugging Face in Brooklyn.": Ini adalah teks di mana model akan mencari entitas berupa nama.

2. grouped_entities=True: Parameter ini mengatur apakah entitas yang ditemukan akan dikelompokkan bersama atau tidak. Dalam hal ini, entitas yang ditemukan akan dikelompokkan berdasarkan jenisnya.

Dengan menggunakan ner yang telah didefinisikan sebelumnya, kita memanggilnya dengan teks yang diberikan sebagai argumen. Hasilnya adalah entitas-entitas berupa nama yang ditemukan dalam teks tersebut, yang dikelompokkan berdasarkan jenisnya.

In [9]:
from transformers import pipeline

question_answerer = pipeline("question-answering")
question_answerer(
    question="Where do I work?",
    context="My name is Sylvain and I work at Hugging Face in Brooklyn",
)

No model was supplied, defaulted to distilbert/distilbert-base-cased-distilled-squad and revision 626af31 (https://huggingface.co/distilbert/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/473 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/261M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

{'score': 0.6949766278266907, 'start': 33, 'end': 45, 'answer': 'Hugging Face'}

Code tersebut menggunakan model Question Answering dari Transformers untuk menjawab pertanyaan yang diajukan berdasarkan konteks yang diberikan. Berikut adalah detailnya:

1. "Where do I work?": Ini adalah pertanyaan yang diajukan.

2. "My name is Sylvain and I work at Hugging Face in Brooklyn": Ini adalah konteks atau teks di mana model akan mencari jawaban atas pertanyaan yang diajukan.

Dengan menggunakan question_answerer yang telah didefinisikan sebelumnya, kita memanggilnya dengan pertanyaan dan konteks yang diberikan sebagai argumen. Hasilnya adalah jawaban atas pertanyaan tersebut yang ditemukan dalam konteks yang diberikan.

In [10]:
from transformers import pipeline

summarizer = pipeline("summarization")
summarizer(
    """
    America has changed dramatically during recent years. Not only has the number of
    graduates in traditional engineering disciplines such as mechanical, civil,
    electrical, chemical, and aeronautical engineering declined, but in most of
    the premier American universities engineering curricula now concentrate on
    and encourage largely the study of engineering science. As a result, there
    are declining offerings in engineering subjects dealing with infrastructure,
    the environment, and related issues, and greater concentration on high
    technology subjects, largely supporting increasingly complex scientific
    developments. While the latter is important, it should not be at the expense
    of more traditional engineering.

    Rapidly developing economies such as China and India, as well as other
    industrial countries in Europe and Asia, continue to encourage and advance
    the teaching of engineering. Both China and India, respectively, graduate
    six and eight times as many traditional engineers as does the United States.
    Other industrial countries at minimum maintain their output, while America
    suffers an increasingly serious decline in the number of engineering graduates
    and a lack of well-educated engineers.
"""
)

No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/1.80k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

[{'summary_text': ' The number of engineering graduates in the United States has declined in recent years . China and India graduate six and eight times as many traditional engineers as the U.S. does . Rapidly developing economies such as China continue to encourage and advance the teaching of engineering . There are declining offerings in engineering subjects dealing with infrastructure, infrastructure, the environment, and related issues .'}]

Dengan menggunakan summarizer yang telah didefinisikan sebelumnya, kita memanggilnya dengan teks yang akan dirangkum sebagai argumen. Hasilnya adalah rangkuman dari teks tersebut.

In [11]:
from transformers import pipeline

translator = pipeline("translation", model="Helsinki-NLP/opus-mt-fr-en")
translator("Ce cours est produit par Hugging Face.")

config.json:   0%|          | 0.00/1.42k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/301M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/293 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/42.0 [00:00<?, ?B/s]

source.spm:   0%|          | 0.00/802k [00:00<?, ?B/s]

target.spm:   0%|          | 0.00/778k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.34M [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/models/marian/tokenization_marian.py:175: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")


[{'translation_text': 'This course is produced by Hugging Face.'}]

Code tersebut menggunakan model terjemahan dari Transformers untuk menerjemahkan teks dari bahasa Prancis ke bahasa Inggris. Berikut adalah detailnya:

1. "Ce cours est produit par Hugging Face.": Ini adalah teks dalam bahasa Prancis yang akan diterjemahkan ke bahasa Inggris. Artinya dalam bahasa Inggris adalah "This course is produced by Hugging Face."

Dengan menggunakan translator yang telah didefinisikan sebelumnya, kita memanggilnya dengan teks dalam bahasa Prancis yang akan diterjemahkan sebagai argumen. Hasilnya adalah teks tersebut yang sudah diterjemahkan ke dalam bahasa Inggris.